src/main/java/zaasclient/HttpsClient.java

```
@@ -0,0 +1,201 @@
+    }
+
+    private CloseableHttpClient getCloseableHttpClient(SSLContext sslContext) {
+        final SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(sslContext,
+            SSLConnectionSocketFactory.getDefaultHostnameVerifier());
+
+        closeableHttpsClient = HttpClients.custom()
+            .setSSLSocketFactory(sslsf)
+            .setDefaultRequestConfig(this.requestConfig)
+            .setMaxConnTotal(3 * 3)
+            .setMaxConnPerRoute(3)
+            .build();
+        return closeableHttpsClient;
+    }
+
+    /**
+     * Overloaded method of the getclosableHttpClient method .. so that the cookies can be used for storing jwt tokens
+     *
+     * @param sslContext
+     * @param cookieStore
+     * @return
+     */
+    private CloseableHttpClient getCloseableHttpClient(SSLContext sslContext, BasicCookieStore cookieStore) {
+        final SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(sslContext,
+            SSLConnectionSocketFactory.getDefaultHostnameVerifier());
+
+        closeableHttpsClient = HttpClients.custom()
+            .setSSLSocketFactory(sslsf)
+            .setDefaultRequestConfig(this.requestConfig)
+            .setMaxConnTotal(3 * 3)
+            .setMaxConnPerRoute(3)
+            .setDefaultCookieStore(cookieStore)
+            .build();
+        return closeableHttpsClient;
+    }
+
+
+    /**
+     * @param kmf
+     * @param tmf
+     * @return
+     * @throws Exception
+     */
+    private SSLContext getSSLContext(KeyManagerFactory kmf, TrustManagerFactory tmf) throws KeyManagementException, NoSuchAlgorithmException {
+        SSLContext sslContext;
+        try {
+            sslContext = SSLContext.getInstance("TLS");
+            sslContext.init(kmf != null ? kmf.getKeyManagers() : null, tmf != null ? tmf.getTrustManagers() : null, new SecureRandom());
+        } catch (NoSuchAlgorithmException | KeyManagementException e) {
+            throw e;
+        }
```

There is no point in catch and immediate rethrow.  The client cannot recover from such error, we can throw the exception out directly. 

We could add a log here (debug) with what was wrong. I would expect the user (developer) is sane enough to catch and log the exception when trying to instantiate our client. Throwed exceptions are declared and the user is responsible to act accordingly.